File: 052_Valid Parentheses.py
"""
Valid Parentheses

Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
determine if the input string is valid (correctly opened and closed).
"""

def is_valid(s: str) -> bool:
    pairs = {')':'(', ']':'[', '}':'{'}
    st = []
    for c in s:
        if c in '([{': st.append(c)
        else:
            if not st or st[-1] != pairs.get(c):
                return False
            st.pop()
    return not st


if __name__ == '__main__':
    print('Par 1:', is_valid('()'))
    print('Par 2:', is_valid('()[]{}'))
    print('Par 3:', is_valid('(]'))

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 052_Valid Parentheses.cpp
/*
Valid Parentheses

Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
determine if the input string is valid (correctly opened and closed).
*/

#include <bits/stdc++.h>
using namespace std;

bool isValid(string s) {
    unordered_map<char,char> m = {{')','('},{']','['},{'}','{'}};
    vector<char> st;
    for (char c: s) {
        if (c == '(' || c == '[' || c == '{') st.push_back(c);
        else {
            if (st.empty() || st.back() != m[c]) return false;
            st.pop_back();
        }
    }
    return st.empty();
}

int main() {
    cout << "Par 1: " << (isValid("()")?"true":"false") << endl;
    cout << "Par 2: " << (isValid("()[]{}")=="true"?"true":"false") << endl; // true
    cout << "Par 3: " << (isValid("(]")?"true":"false") << endl;
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)
